From c1261cce0fa0c0fad87327995f980c6aec564c98 Mon Sep 17 00:00:00 2001
From: Philipp Zabel <p.zabel@pengutronix.de>
Date: Fri, 23 Aug 2019 14:37:00 +0200
Subject: [PATCH 02/13] v4l2: introduce v4l2_get_controls

This can be used to query codec mode controls,
such as decode mode and start code for h.264.

Signed-off-by: Philipp Zabel <p.zabel@pengutronix.de>
---
 src/v4l2.c | 35 ++++++++++++++++++++++++++++++-----
 src/v4l2.h |  3 +++
 2 files changed, 33 insertions(+), 5 deletions(-)

diff --git a/src/v4l2.c b/src/v4l2.c
index 352047b..3addb33 100644
--- a/src/v4l2.c
+++ b/src/v4l2.c
@@ -428,12 +428,11 @@ int v4l2_export_buffer(int video_fd, unsigned int type, unsigned int index,
 	return 0;
 }
 
-int v4l2_set_controls(int video_fd, int request_fd,
-		      struct v4l2_ext_control *control_array,
-		      unsigned int num_controls)
+static int v4l2_ioctl_controls(int video_fd, int request_fd, unsigned long ioc,
+			       struct v4l2_ext_control *control_array,
+			       unsigned int num_controls)
 {
 	struct v4l2_ext_controls controls;
-	int rc;
 
 	memset(&controls, 0, sizeof(controls));
 
@@ -445,7 +444,33 @@ int v4l2_set_controls(int video_fd, int request_fd,
 		controls.request_fd = request_fd;
 	}
 
-	rc = ioctl(video_fd, VIDIOC_S_EXT_CTRLS, &controls);
+	return ioctl(video_fd, ioc, &controls);
+}
+
+int v4l2_get_controls(int video_fd, int request_fd,
+		      struct v4l2_ext_control *control_array,
+		      unsigned int num_controls)
+{
+	int rc;
+
+	rc = v4l2_ioctl_controls(video_fd, request_fd, VIDIOC_G_EXT_CTRLS,
+				 control_array, num_controls);
+	if (rc < 0) {
+		request_log("Unable to get control(s): %s\n", strerror(errno));
+		return -1;
+	}
+
+	return 0;
+}
+
+int v4l2_set_controls(int video_fd, int request_fd,
+		      struct v4l2_ext_control *control_array,
+		      unsigned int num_controls)
+{
+	int rc;
+
+	rc = v4l2_ioctl_controls(video_fd, request_fd, VIDIOC_S_EXT_CTRLS,
+				 control_array, num_controls);
 	if (rc < 0) {
 		request_log("Unable to set control(s): %s\n", strerror(errno));
 		return -1;
diff --git a/src/v4l2.h b/src/v4l2.h
index 3bccb23..24c12a0 100644
--- a/src/v4l2.h
+++ b/src/v4l2.h
@@ -54,6 +54,9 @@ int v4l2_dequeue_buffer(int video_fd, int request_fd, unsigned int type,
 int v4l2_export_buffer(int video_fd, unsigned int type, unsigned int index,
 		       unsigned int flags, int *export_fds,
 		       unsigned int export_fds_count);
+int v4l2_get_controls(int video_fd, int request_fd,
+		      struct v4l2_ext_control *controls,
+		      unsigned int num_controls);
 int v4l2_set_controls(int video_fd, int request_fd,
 		      struct v4l2_ext_control *controls,
 		      unsigned int num_controls);
-- 
2.30.2

